fdk/http/routine.py

```diff
@@ -0,0 +1,157 @@
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+import asyncio
+import h11
+import io
+import traceback
+import sys
+
+from fdk import constants
+from fdk import log
+from fdk import response
+
+
+def ensure_connection_is_safe_to_write(action):
+    async def inner(*args, **kwargs):
+        try:
+            return await action(*args, **kwargs)
+        except h11.ProtocolError as ex:
+            log.log(str(ex))
+            traceback.print_exc(file=sys.stderr)
+            sys.exit("Fn <-> FDK connectivity issue: ".format(str(ex)))
+
+    return inner
+
+
+async def process_chunk(connection: h11.Connection,
+                        request_reader: asyncio.StreamReader):
+    """
+    Reads the request in chunks until can parse
+    the request successfully
+    :param connection: h11 request parser
+    :type connection: h11.Connection
+    :param request_reader: async stream reader
+    :type request_reader: asyncio.StreamReader
+    :return: request and body
+    :rtype tuple
+    """
+    # TODO(denismakogon): replace io.BytesIO with asyncio.StreamReader
+    # this change will be required when an FDK will enforce customer's
+    # function to be a coroutine
+    request, body = None, io.BytesIO()
+    log.log("starting process_chunk")
+    while True:
+        log.log("process_chunk: reading chunk of data from async reader")
+        buf = await request_reader.read(constants.ASYNC_IO_READ_BUFFER)
+        log.log("process_chunk: buffer filled")
+        connection.receive_data(buf)
+        log.log("process_chunk: sending data to h11")
+        while True:
+            event = connection.next_event()
+            log.log("process_chunk: event type {0}"
+                    .format(type(event)))
+            if isinstance(event, h11.Request):
+                request = event
+            if isinstance(event, h11.Data):
+                body.write(event.data)
+            if isinstance(event, h11.EndOfMessage):
+                return request, body
+            if isinstance(event, (h11.NEED_DATA, h11.PAUSED)):
```

Comment:
we need to handle NEED_DATA and do the actual `connection.receive_data` from here in this case for NEED_DATA only (not PAUSED too), we can't just read from the socket as it may eg be a MUST_CLOSE event